rebuild/src/signature.cc

```diff
@@ -0,0 +1,166 @@
+    // Setup object correctly.  Start with type tagging wrapper class.
+    wrapped.TypeTag(&module->_signature_tag);
+    // Unwrap object to get native instance
+    Signature *sig = Signature::Unwrap(wrapped);
+    // default to jacobian for now
+    sig->_has_jacobian = true;
+
+    // but figure out if request for affine
+    if (!info[1].IsUndefined())
+    {
+        Napi::Value type_val = info[1].As<Napi::Value>();
+        if (!type_val.IsNumber())
+        {
+            Napi::TypeError::New(env, "type must be of enum CoordType (number)").ThrowAsJavaScriptException();
+            return scope.Escape(env.Undefined());
+        }
+        if (type_val.As<Napi::Number>().Uint32Value() == 0)
+        {
+            sig->_has_jacobian = false;
+            sig->_has_affine = true;
+        }
+    }
+
+    try
+    {
+        if (sig->_has_jacobian)
+        {
+            sig->_jacobian.reset(new blst::P2{sig_bytes.Data(), sig_bytes.ByteLength()});
+        }
+        else
+        {
+            sig->_affine.reset(new blst::P2_Affine{sig_bytes.Data(), sig_bytes.ByteLength()});
+        }
+    }
+    catch (blst::BLST_ERROR err)
+    {
+        Napi::RangeError::New(env, module->GetBlstErrorString(err)).ThrowAsJavaScriptException();
+        return scope.Escape(env.Undefined());
+    }
+
+    return scope.Escape(wrapped);
+}
+
+Signature::Signature(const Napi::CallbackInfo &info)
+    : Napi::ObjectWrap<Signature>{info},
+      _has_jacobian{false},
+      _has_affine{false},
+      _jacobian{nullptr},
+      _affine{nullptr}
+{
+    Napi::Env env = info.Env();
+    Napi::HandleScope scope(env);
+    // Check that constructor was called from C++ and not JS. Externals can only
+    // be created natively.
+    if (!info[0].IsExternal())
+    {
+        Napi::Error::New(env, "Signature constructor is private").ThrowAsJavaScriptException();
+        return;
+    }
+};
+
+Napi::Value Signature::Serialize(const Napi::CallbackInfo &info)
+{
+    Napi::Env env = info.Env();
+    Napi::EscapableHandleScope scope(env);
+
+    bool compressed{true};
+    if (!info[0].IsUndefined())
+    {
+        compressed = info[0].ToBoolean().Value();
+    }
+    Napi::Buffer<uint8_t> serialized = Napi::Buffer<uint8_t>::New(
+        env,
+        compressed
+            ? BLST_TS_SIGNATURE_LENGTH_COMPRESSED
+            : BLST_TS_SIGNATURE_LENGTH_UNCOMPRESSED);
+
+    if (_has_jacobian)
+    {
+        compressed ? _jacobian->compress(serialized.Data()) : _jacobian->serialize(serialized.Data());
+    }
+    else if (_has_affine)
+    {
+        compressed ? _affine->compress(serialized.Data()) : _affine->serialize(serialized.Data());
+    }
+    else
+    {
+        Napi::Error::New(env, "Signature cannot be serialized. No point found!").ThrowAsJavaScriptException();
+        return scope.Escape(env.Undefined());
+    }
+
+    return scope.Escape(serialized);
+}
+
+Napi::Value Signature::SigValidate(const Napi::CallbackInfo &info)
```

Comment:
i think this validate fn seems to be same for public key and sig, so may be could be turned into a helper/macro call?

Reply:
There is the added `is_inf` check for P1 that is not in P2 validation.  Only pk's need to be verified as not the identity point.

Parameters:
https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-2.2

Key Validate:
https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-2.5